lib/luaxtable.c

```diff
@@ -0,0 +1,234 @@
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/module.h>
+#include <linux/version.h>
+#include <linux/types.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter/x_tables.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+typedef enum luaxtable_type_e {
+	LUAXTABLE_TMATCH,
+	LUAXTABLE_TTARGET,
+} luaxtable_type_t;
+
+typedef struct luaxtable_s {
+	lunatik_object_t *runtime;
+	union {
+		struct xt_match *match;
+		struct xt_target *target;
+	};
+	luaxtable_type_t type;
+} luaxtable_t;
+
+static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+{
+	return true;
+}
+
+static int luaxtable_match_check(const struct xt_mtchk_param *par)
+{
+	return 0;
+}
+
+static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+{
+}
+
+static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+{
+	return NF_ACCEPT;
+}
+
+static int luaxtable_target_check(const struct xt_tgchk_param *par)
+{
+	return 0;
+}
+
+static void luaxtable_target_destroy(const struct xt_tgdtor_param *par)
+{
+}
+
+static void luaxtable_release(void *private);
+
+static const luaL_Reg luaxtable_mt[] = {
+	{"__gc", lunatik_deleteobject},
+	{NULL, NULL}
+};
+
+static const lunatik_class_t luaxtable_class = {
+	.name = "xtable",
+	.methods = luaxtable_mt,
+	.release = luaxtable_release,
+	.sleep = false,
+};
+
+#define luaxtable_setinteger(L, idx, hook, field) 		\
+do {								\
+	lunatik_checkfield(L, idx, #field, LUA_TNUMBER);	\
+	hook->field = lua_tointeger(L, -1);			\
+	lua_pop(L, 1);						\
+} while (0)
+
+#define luaxtable_setstring(L, idx, hook, field, maxlen)        \
+do {								\
+	const char *str;					\
+	size_t len;						\
+								\
+	lunatik_checkfield(L, idx, #field, LUA_TSTRING);	\
+	str = lua_tolstring(L, -1, &len);			\
+	if (len >= maxlen)					\
+		luaL_error(L, "'%s' is too long", #field);	\
+	strncpy(hook->field, str, maxlen);			\
+	lua_pop(L, 1);						\
+} while (0)
+
+static inline lunatik_object_t *luaxtable_new(lua_State *L, int idx, int hook)
+{
+	lunatik_object_t *object = lunatik_newobject(L, &luaxtable_class , sizeof(luaxtable_t));
+	luaxtable_t *xtable = (luaxtable_t *)object->private;
+
+	xtable->runtime = lunatik_toruntime(L);
+	xtable->type = hook;
+	lunatik_getobject(xtable->runtime);
+	lunatik_registerobject(L, idx, object);
+	return object;
+}
+
+#define LUAXTABLE_NEWHOOK(hook, HOOK)					\
+static int luaxtable_new##hook(lua_State *L) 				\
+{									\
+	luaL_checktype(L, 1, LUA_TTABLE);				\
+									\
+	lunatik_object_t *object = luaxtable_new(L, 1, HOOK); 		\
```

Comment:
I don't think we should register the object here; if we raise an error afterwards, the object won't be collected because we hold a reference.. registering the object should be the last thing to do on this macro.. we probably need to split (or get rid of) this luaxtable_new() function..

Reply:
btw, it would be good to run some tests forcing the error condition, putting some prints on release and also reloading lunatik (to observe if it will hang)..